File: LeetCode_ProblemSet/Problem3394.cpp
#include <algorithm>
#include <vector>
#include <functional>

class Solution {
   public:
    bool checkValidCuts(int n, std::vector<std::vector<int>>& rectangles) {
        for (int i = 0; i < 2; i++) {
            std::vector<std::pair<int, int>> coords;
            coords.reserve(rectangles.size());

            for (const auto& r : rectangles) {
                coords.emplace_back(r[i], r[i + 2]);
            }

            std::sort(coords.begin(), coords.end());

            int cnt = 0, cur = 0;
            for (const auto& p : coords) {
                int start = p.first, end = p.second;

                if (start < cur) {
                    cur = std::max(cur, end);
                    continue;
                }
                cnt++;
                if (cnt == 3) return true;
                cur = end;
            }
        }

        return false;
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem3394.rs
use std::cmp::max;

impl Solution {
    pub fn check_valid_cuts(_: i32, rectangles: Vec<Vec<i32>>) -> bool {
        let mut x_coords: Vec<(i32, i32)> = rectangles.iter().map(|r| (r[0], r[2])).collect();
        let mut y_coords: Vec<(i32, i32)> = rectangles.iter().map(|r| (r[1], r[3])).collect();

        x_coords.sort();
        y_coords.sort();

        for coords in vec![x_coords, y_coords] {
            let (mut cnt, mut cur) = (0, 0);
            for (start, end) in coords {
                if start < cur {
                    cur = max(cur, end);
                    continue;
                } 
                cnt += 1;
                if cnt == 3 {
                    return true;
                }
                cur = end;
            }
        }

        return false;
    }
}